crates/matrix-sdk-crypto/src/machine/mod.rs

```diff
@@ -1465,11 +1462,32 @@ impl OlmMachine {
                 // No - use the existing data.
                 session.sender_data.clone()
             }
+        } else {
+            session.sender_data.clone()
         };
 
         Ok(sender_data_to_verification_state(sender_data, session.has_been_imported()))
     }
 
+    /// Whether we should recalculate the Megolm sender's data, given the
+    /// current sender data. We only want to recalculate if it might increase
+    /// trust and allow us to decrypt messages that we otherwise might
+    /// refuse to decrypt.
+    ///
+    /// We recalculate for all states except:
+    ///
+    /// - SenderUnverified: the sender is trusted enough that we will decrypt
+    ///   their messages in all cases, or
+    /// - SenderVerified: the sender is the most trusted they can be.
+    fn should_recalculate_sender_data(sender_data: &SenderData) -> bool {
```

Comment:
I don't really get why you want to move this to `OlmMachine`. `OlmMachine` is way too big as it is, and keeping it as a method on `SenderData` (with another name, maybe) seems preferable.

Reply:
When it was on `SenderData` it appeared to be a general-purpose method. What we discovered as we discussed it was that the logic here is tightly coupled to when we want to recalculate sender data when we decrypt a message. It will change if we change our minds (e.g. to recalculate when we have an unverified sender in case they became verified).

So it feels like an implementation detail of `get_or_update_verification_state`. I'd be happy to move it literally inside `get_or_update_verification_state` if you like.

Reply:
ok possibly this is a sign that the whole of `get_or_update_verification_state`  needs to move somewhere else.

Please either make it an inner function or give it a name and doc-comment which identify it is part of `get_or_update_verification_state`. I don't have a strong feeling which.

Reply:
Moved to an inner function in 3a84196573